### ISSUE TITLE
core: Could Vault manage known config files?

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/vault#1246]
The standard day-to-day workflow for Vault user is the following:
1. `vault auth -method=abc username=johndoe` => ends up with a valid `~/.vault-token`
2. `vault read aws/creds/role-name`
3. Copy the credentials from the screen
4. Open `~/.aws/credentials` and paste the credentials + tune the format so that AWS CLI is happy

There is a lot of room for error when copy-pasting + it will soon become bothersome, especially with keys that have been leased for short time period.

It's not only common workflow for AWS, but for many other secret backends. There are standardised locations & formats of config files for many:
- AWS - `~/.aws/credentials` or file with standard ENV vars to be `source`d
- Docker - `~/.docker/config.json`
- [PostgreSQL](http://www.postgresql.org/docs/9.2/static/libpq-pgpass.html) - `~/.pgpass`
- [MySQL](http://dev.mysql.com/doc/refman/5.7/en/option-files.html) - `~/.my.cnf`
- SSH - `~/.ssh/`
- etc.

Would there be any interest in getting the logic for updating config files into Vault?

**Comment by jefferai** [comment]
Hi @radeksimko ,

I think it's interesting if it could be done in a fairly fool-proof way, and probably couldn't be done by default. Keep in mind, however, that the Vault CLI is just a wrapper around the JSON API, so it's already possible to use a minimal script or a tool like `jq` to pull pieces of the response out and populate configuration files with them, and it's quite possible that an external tool is the right way to go here.

**Comment by DylanFrese** [comment]
I don't think that Vault is the tool for this. I'm in the middle of writing a templating system that generates configuration frameworks for different servers in different environments and regions, and queries Vault to get secrets for that particular environment. My team uses Docker, so it makes sense to build configuration this way, but in a more traditional setup, Configuration/Management software could accomplish this. SaltStack, for example, is looking to support Vault as an external pillar database (https://github.com/saltstack/salt/issues/27020).

Basically, the problem is that if you're populating the configuration by hand already, you should be using configuration/management regardless of whether you're pulling from Vault or not. If you're already using configuration/management, then that is what should be querying Vault to write the configuration file.

**Comment by radeksimko** [comment]
@DylanFrese Totally agreed about cfg mgmt for **machines**. I think you missed the original point though. I was talking about secrets for humans/users, not machines. These are two different use cases.

**Comment by mtougeron** [comment]
+1 for this request. We would like to use Vault with STS AssumeRole so that we don't need to manage AWS access keys/secrets via the IAM user accounts (or DirectoryService users). This way once a user has been removed from AD they won't be able to access the AWS Console nor will they be able to authenticate to Vault to get the AWS secrets.

But the difficulty at this point is that we either need to get wrapper script(s) onto everyone's computers or rely on users to properly copy/paste the Vault output into their `~/.aws/credentials` file in order to use the AWS CLI. If Vault could take the output from the `vault read aws/creds/role-name` and put it directly into the `~/.aws/credentials` file (maybe under the profile that matches the `role-name`?) it would save us a lot of headaches.

It should also probably respect the `AWS_CONFIG_FILE` environment variable but I defer to you on that.

**Comment by jefferai** [comment]
Right now the normal workflow for this would be to use Consul-Template.

**Comment by mtougeron** [comment]
@jefferai That would work for machine based credentials but not necessarily for real people. We wouldn't want it to manipulate anything else in the `~/.aws/credentials` file other than the specific set of aws credentials. e.g., if the user has modified their default `region`, `s3` or any other setting then `consul-template` would stomp over those settings.

We would also need to maintain and figure out an update mechanism for the template files on each user's computer. If we have to do that, then we might as well use a wrapper script.

**Comment by mtougeron** [comment]
BTW, I'm thinking this type of functionality would be just for the dynamic credential types that Vault supports. Kind of like how it has the wrapper for `vault ssh` to use those dynamic credentials.

**Comment by jefferai** [comment]
We don't currently have plans to implement this, so closing.

**Comment by radeksimko** [comment]
Just wanted to note that many of the use cases which originally motivated the creation of this issue seem to be solved by Boundary, which was just released 😄 

https://www.boundaryproject.io

or to be more specific Boundary seems to make a lot of these configs practically unnecessary as credentials (if at all needed) are short-lived for the duration of the session and in many cases never persisted
